node(transfer verifier): Update TransferIsValid method to return results per Message rather than per Receipt

johnsaigle opened the pull request:
- Update TransferIsValid method to support per-message validation

- Add `msgID` parameter (equivalent to VAA ID) to `TransferIsValid` method
  signature to enable validation of specific messages within transaction
  receipts. Update caching mechanism to store per-message validation
  results instead of per-transaction results. Modify receipt processing
  to track message-specific evaluations and validate transfer amounts on
  a per-message basis.

Key changes:
- Add msgID parameter to TransferIsValid method for targeted message validation
- Replace single Result field in evaluation struct with Results map keyed by message ID
- Update receipt summary to track transfers per message ID rather than aggregated amounts
- Modify invariant checking to validate each message's transfers individually
- Add MsgID method to generate unique message identifiers from LogMessagePublished events
- Add sequence field to LogMessagePublished struct for message identification

pleasew8t commented on node/pkg/txverifier/evm.go:
```diff
@@ -40,20 +41,24 @@ func (i InvariantError) Error() string {
 // Return values:
 //
-//	true:  the token transfer's receipt is valid.
+//	true: the token transfer's in the messagepublication is valid.
 //	false and nil: the token transfer has violated an invariant and is unsafe.
```
also, `messagepublication` -> `message publication`

johnsaigle replied:
Yeah I kind of hate `bool, err` as a return type in Go. We could return a custom struct or something, but it feels important to distinguish between parsing and invariant errors, idk.
`true, err` should never be possible but it's worth making that super clear.

johnsaigle replied:
I've added more comments and rewritten (hopefully: simplified) the return values of this function. I believe it's impossible to hit `true, err` now but let me know if I made a mistake.

pleasew8t commented on node/pkg/txverifier/evm.go:
```diff
@@ -40,20 +41,24 @@ func (i InvariantError) Error() string {
 //	false and err: the receipt could not be properly processed or is not a token transfer.
 func (tv *TransferVerifier[ethClient, Connector]) TransferIsValid(
 	ctx context.Context,
+	// The message ID is a combination of the emitter chain, emitter address, and sequence.
```
nit: we could add here that it's similar to the VAA ID that would eventually get generated? That way there's no confusion about what this is.

johnsaigle replied:
Yeah I could add something like "See the SDK" that has MessageID function that produces the same output

pleasew8t commented on node/pkg/txverifier/evm.go:
```diff
@@ -123,33 +150,36 @@ func (tv *TransferVerifier[ethClient, Connector]) TransferIsValid(
 
-	if processErr != nil {
-		// Check if the error type is an invariant error. If not, it's just a parsing error.
+	if validateReceiptErr != nil {
+		// Check if the error type is an invariant error. If not, it's just a parsing error. It's important to
```
I'm not entirely sure how this responds differently to a single invalid transfer in the batch invalidating the entire transaction. I think the flow needs to change here:
* check if a non-invariant-violation error ocurred, and return `false, err` if so
* if the errors contain an invariant violation && the msg ID is empty, return `false, nil`

Basically, if a MsgId did exist, we continue with caching the evaluations, and return from the cached evaluation the result for the specific MsgId.

johnsaigle replied:
Hey I rewrote this section with the idea that it could be easier to follow. Let me know what you think about the new version.

pleasew8t commented on node/pkg/txverifier/evmtypes.go:
```diff
@@ -506,19 +535,22 @@ func parseERC20TransferEvent(logTopics []common.Hash, logData []byte) (common.Ad
-	// Which contract emitted the event.
+	// Which contract emitted the event. (not to be confused with EmitterAddress in a VAA)
 	EventEmitter common.Address
 	// Which address sent the transaction that triggered the message publication.
 	MsgSender common.Address
```
nit: "The address of the account that initiated the bridge transfer."

pleasew8t commented on node/pkg/txverifier/evmtypes.go:
```diff
@@ -1002,6 +1041,11 @@ func validate[L TransferLog](tLog TransferLog) error {
 			return &InvalidLogError{Msg: "destination is not set"}
 		}
 
+		// It's invalid for the core bridge to send a message to itself.
+		if Cmp(log.Sender(), log.Emitter()) == 0 {
```
Honestly, I'd say if this happens it's crazy suspicious and not just an invalid error. Something to alert on perhaps?

pleasew8t commented on node/pkg/txverifier/evmtypes_test.go:
```diff
@@ -532,6 +529,19 @@ func TestValidateLogMessagePublished(t *testing.T) {
 				},
 			},
 		},
+		"invalid: msg.sender cannot be equal to emitter": {
```
Shouldn't this be "emitter cannot be core bridge"? Remember to remove if you decide to remove the core bridge check.

johnsaigle replied:
I think the test checks any sender against any emitter rather than checking emitter against the hard-coded core bridge address, so I think the existing name is more accurate. Do you feel otherwise?

johnsaigle commented on node/pkg/txverifier/evmtypes.go:
```diff
@@ -639,42 +673,56 @@ func (r *TransferReceipt) String() string {
 
 // Summary of a processed TransferReceipt. Contains information about relevant
-// transfers requested in and out of the bridge.
+// transfers requested in and out of the bridge. Message Publications within a receipt
```
I'm going to leave this as-is because it refers to a specific type `MessagePublication` that is used elsewhere in the Guardian

djb15 commented on node/pkg/txverifier/evm_test.go:
```diff
@@ -672,8 +911,190 @@ func TestValidateReceipt(t *testing.T) {
+			// Parsing error, so the summary should be nil.
+			// expected: ReceiptSummary{},
+		},
+		"unsafe receipt: total amount in is too small even though messages are valid when considered individually": {
```
This test case is targeting the wrong thing I think; the input tokens are native, output are USDC

djb15 replied:
Based on my earlier comments I can get this test case to pass when it shouldn't after correcting the tokens with the following diff. We deposit 3 tokens and withdraw 4, and no errors are thrown.

```
diff --git a/node/pkg/txverifier/evm_test.go b/node/pkg/txverifier/evm_test.go
index f10e3be63..36498a24c 100644
--- a/node/pkg/txverifier/evm_test.go
+++ b/node/pkg/txverifier/evm_test.go
@@ -974,8 +974,8 @@ func TestValidateReceipt(t *testing.T) {
 						TokenChain:   vaa.ChainIDEthereum,
 						From:         eoaAddrGeth,
 						To:           tokenBridgeAddr,
-						Amount:       big.NewInt(1),
-						OriginAddr:   nativeAddrVAA,
+						Amount:       big.NewInt(3),
+						OriginAddr:   usdcAddrVAA,
 					},
 				},
 				MessagePublications: &[]*LogMessagePublished{
@@ -985,10 +985,10 @@ func TestValidateReceipt(t *testing.T) {
 						Sequence:     1,
 						TransferDetails: &TransferDetails{
 							PayloadType:   TransferTokens,
-							OriginAddress: nativeAddrVAA,
+							OriginAddress: usdcAddrVAA,
 							TargetAddress: eoaAddrVAA,
 							TokenChain:    2,
-							Amount:        big.NewInt(1),
+							Amount:        big.NewInt(2),
 						},
 					},
 					{
@@ -997,37 +997,27 @@ func TestValidateReceipt(t *testing.T) {
 						Sequence:     2,
 						TransferDetails: &TransferDetails{
 							PayloadType:   TransferTokens,
-							OriginAddress: nativeAddrVAA,
+							OriginAddress: usdcAddrVAA,
 							TargetAddress: eoaAddrVAA,
 							TokenChain:    2,
-							Amount:        big.NewInt(1),
+							Amount:        big.NewInt(2),
 						},
 					},
 				},
 			},
 			expected: ReceiptSummary{
 				in: map[string]*big.Int{
-					fmt.Sprintf("%s-%s", nativeAddrVAA, "2"): big.NewInt(1),
+					fmt.Sprintf("%s-%s", usdcAddrVAA, "2"): big.NewInt(1),
 				},
 				out: map[msgID]transferOut{
-					{Sequence: 1, EmitterChain: vaa.ChainIDEthereum, EmitterAddress: VAAAddrFrom(tokenBridgeAddr)}: {fmt.Sprintf("%s-%s", nativeAddrVAA, "2"), big.NewInt(1)},
-					{Sequence: 2, EmitterChain: vaa.ChainIDEthereum, EmitterAddress: VAAAddrFrom(tokenBridgeAddr)}: {fmt.Sprintf("%s-%s", nativeAddrVAA, "2"), big.NewInt(1)},
+					{Sequence: 1, EmitterChain: vaa.ChainIDEthereum, EmitterAddress: VAAAddrFrom(tokenBridgeAddr)}: {fmt.Sprintf("%s-%s", usdcAddrGeth, "2"), big.NewInt(1)},
+					{Sequence: 2, EmitterChain: vaa.ChainIDEthereum, EmitterAddress: VAAAddrFrom(tokenBridgeAddr)}: {fmt.Sprintf("%s-%s", usdcAddrGeth, "2"), big.NewInt(1)},
 				},
 				msgPubResult: map[msgID]bool{
 					{Sequence: 1, EmitterChain: vaa.ChainIDEthereum, EmitterAddress: VAAAddrFrom(tokenBridgeAddr)}: true,
 					{Sequence: 2, EmitterChain: vaa.ChainIDEthereum, EmitterAddress: VAAAddrFrom(tokenBridgeAddr)}: true,
 				},
-				problems: &Problems{
-					// This manifests as two problems because each individual message out is ruled insolvent relative to the whole receipt.
-					&ReceiptProblem{
-						violationKind: ReceiptInvariant,
-						err:           *ErrInvariantReceiptInsolvent,
-					},
-					&ReceiptProblem{
-						violationKind: ReceiptInvariant,
-						err:           *ErrInvariantReceiptInsolvent,
-					},
-				},
+				problems: &Problems{},
 			},
 		},
 		"unsafe receipt: one valid message, one invalid message": {

```

johnsaigle replied:
Thanks I've added an adapted version of this test in the latest commit.

johnsaigle commented on node/pkg/txverifier/evm.go:
```diff
@@ -705,62 +749,132 @@ func (tv *TransferVerifier[evmClient, connector]) validateReceipt(
+
+// isSolvent() reports whether the sum of the outbound transfers is at least as much as the sum of the inbound transfers.
+// It is used to examine the solvency of the entire receipt, without considering individual messages.
+func (receiptSummary *ReceiptSummary) isSolvent() bool {
```
I added the state check in this function in the latest commit

johnsaigle commented:
@djb15 I removed the Problem struct in the latest commit. I felt like it was possible to get the same thing done even without a boolead, instead relying on the the `isSafe()` and `isMsgSafe()` methods. Let me know what you think.

johnsaigle commented:
@djb15 I removed the else block you flagged.

I noticed after this though that some msgPubs were missing in the unit tests. It turns out that this loop was the part that ensured each message publication has a `msgPub` entry in the summary.

I changed this so that, as the LogMessagePublished parts are being parsed/validated, it also adds msgPub entry to the the map.

It has a default entry of `true` which the insolvency and missing assets checks then can mark as false. Another way to do this would be to start everything as `false`, then later after all the checks are done, go back and find the _solvent_ assets, and mark those as true.

pleasew8t commented on node/pkg/txverifier/evm.go:
```diff
@@ -69,11 +70,41 @@ func (tv *TransferVerifier[ethClient, Connector]) TransferIsValid(
+	}
+
+	// Check to see if the receipt is cached.
+	if receiptEvaluation, exists := tv.evaluations[txHash]; exists {
```
or `rcptEvaluation`

johnsaigle commented on node/pkg/txverifier/evm.go:
```diff
@@ -612,7 +651,7 @@ func (tv *TransferVerifier[evmClient, connector]) validateReceipt(
 	}
 
-	invalidErr := receipt.Validate()
+	invalidErr := receipt.SanityCheck()
 	if invalidErr != nil {
```
Maybe we could save optimizing the code for another follow-up PR? I'm happy if you want to collect this and other feedback in a GH issue that we can come back to.

pleasew8t commented on node/pkg/txverifier/utils.go:
```diff
@@ -26,6 +27,69 @@ const (
+
+	supported := IsSupported(chainID)
+	if !supported {
+		return msgID{}, err
```
If we return here, the error is going to be `nil`. Is this intentional (return empty msgID + nil error), or did you want to return an actual error here?

johnsaigle replied:
nice catch, no this should be an error, not nil.

evan-gray reviewed (approved):
only reviewed code owner changes in watchers/evm
